File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
from typing import List
class Solution:
    
    def solve(self , nums: List[int] , target : int , i : int , n : int , dp : List[List[int]]) -> bool:
        if(target == 0):
            return True
        
        if(target < 0):
            return False
        
        if(i == n):
            return target == 0
        
        if(dp[i][target] != -1):
             return dp[i][target]

        take = self.solve(nums , target - nums[i] , i + 1 , n , dp)
        skip = self.solve(nums , target , i + 1 , n , dp)

        dp[i][target] = (take or skip)
        return dp[i][target]

    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        target = total // 2
        if(total % 2 == 1):
            return False
        
        dp = [[-1 for _ in range(target + 1)] for _ in range(n)]

        return self.solve(nums , target , 0 , n , dp)

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:

    int n;

    vector<vector<int>>dp;
    
    int solve(vector<int>& arr , int sum , int i){
        if(sum == 0) return 1;
        if(i >= n) return sum == 0;
        if(sum < 0) return 0;
        if(dp[i][sum] != -1) return dp[i][sum];
        int skip = solve(arr , sum , i + 1);
        int take = 0;
        
        if(arr[i] <= sum){
            take = solve(arr , sum - arr[i] , i + 1);
        }
        
        return dp[i][sum] = take || skip;
    }

    bool canPartition(vector<int>& nums) {
        n = nums.size();
        int sum = accumulate(nums.begin() , nums.end() , 0);
        if(sum % 2 != 0) return false;
        int target = sum / 2;
        dp.assign(n , vector<int>(target + 1 , -1));
        return solve(nums , target , 0);
    }
};